compiler/src/dotty/tools/scripting/ScriptingDriver.scala

```diff
@@ -0,0 +1,79 @@
+package dotty.tools.scripting
+
+import java.nio.file.{ Files, Path }
+import java.io.File
+import java.net.{ URL, URLClassLoader }
+import java.lang.reflect.{ Modifier, Method }
+
+import scala.jdk.CollectionConverters._
+
+import dotty.tools.dotc.{ Driver, Compiler }
+import dotty.tools.dotc.core.Contexts, Contexts.{ Context, ContextBase, ctx }
+import dotty.tools.dotc.config.CompilerCommand
+import dotty.tools.io.{ PlainDirectory, Directory }
+import dotty.tools.dotc.reporting.Reporter
+import dotty.tools.dotc.config.Settings.Setting._
+
+import sys.process._
+
+class ScriptingDriver(compilerArgs: Array[String], scriptFile: File, scriptArgs: Array[String]) extends Driver:
+  def compileAndRun(): Unit =
+    val outDir = Files.createTempDirectory("scala3-scripting")
+    val (toCompile, rootCtx) = setup(compilerArgs :+ scriptFile.getAbsolutePath, initCtx.fresh)
+    given Context = rootCtx.fresh.setSetting(rootCtx.settings.outputDir,
+      new PlainDirectory(Directory(outDir)))
+
+    if doCompile(newCompiler, toCompile).hasErrors then
+      throw ScriptingException("Errors encountered during compilation")
+
+    try detectMainMethod(outDir, ctx.settings.classpath.value).invoke(null, scriptArgs)
+    catch
+      case e: java.lang.reflect.InvocationTargetException =>
+        throw e.getCause
+
+  end compileAndRun
+
+  private def detectMainMethod(outDir: Path, classpath: String): Method =
+    val outDirURL = outDir.toUri.toURL
+    val classpathUrls = classpath.split(":").map(File(_).toURI.toURL)
+    val cl = URLClassLoader(classpathUrls :+ outDirURL)
+
+    def collectMainMethods(target: File, path: String): List[Method] =
+      val nameWithoutExtension = target.getName.takeWhile(_ != '.')
+      val targetPath =
+        if path.nonEmpty then s"${path}.${nameWithoutExtension}"
+        else nameWithoutExtension
+
+      if target.isDirectory then
+        for
+          packageMember <- target.listFiles.toList
+          membersMainMethod <- collectMainMethods(packageMember, targetPath)
+        yield membersMainMethod
+      else if target.getName.endsWith(".class") then
+        val cls = cl.loadClass(targetPath)
+        val method =
+          try cls.getMethod("main", classOf[Array[String]])
+          catch
+            case _: java.lang.NoSuchMethodException => null
+
+        if method != null && Modifier.isStatic(method.getModifiers) then List(method)
+        else Nil
+      else Nil
+    end collectMainMethods
+
+    val candidates = for
+      file <- outDir.toFile.listFiles.toList
```

Comment:
```suggestion
      file <- outDir.toFile.listFiles.toIterator
```
or some other way to get the iterator from java to scala